modules/csvparser/filterx-func-format-csv.c

```diff
@@ -0,0 +1,255 @@
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "filterx-func-format-csv.h"
+#include "filterx/expr-literal.h"
+#include "filterx/object-string.h"
+#include "filterx/object-null.h"
+#include "filterx/object-dict-interface.h"
+#include "filterx/object-list-interface.h"
+#include "filterx/filterx-eval.h"
+
+#include "scratch-buffers.h"
+#include "utf8utils.h"
+
+#define FILTERX_FUNC_FORMAT_CSV_USAGE "Usage: format_csv(csv_dict or csv_array)"
+
+typedef struct FilterXFunctionFormatCSV_
+{
+  FilterXFunction super;
+  FilterXExpr *input;
+  gchar delimiter;
+  FilterXExpr *columns;
+} FilterXFunctionFormatCSV;
+
+static gboolean
+_append_to_buffer(FilterXObject *key, FilterXObject *value, gpointer user_data)
+{
+  if (!value)
+    return FALSE;
+
+  FilterXFunctionFormatCSV *self = ((gpointer *) user_data)[0];
+  GString *buffer = ((gpointer *) user_data)[1];
+
+  if (filterx_object_is_type(value, &FILTERX_TYPE_NAME(dict)) ||
+      filterx_object_is_type(value, &FILTERX_TYPE_NAME(list)))
+    {
+      msg_debug("FilterX: format_csv(): skipping object, type not supported",
+                evt_tag_str("type", value->type->name));
+      return TRUE;
+    }
+
+  if (buffer->len)
+    g_string_append(buffer, &self->delimiter);
+
+  gsize len_before_value = buffer->len;
+  if (!filterx_object_repr_append(value, buffer))
+    return FALSE;
+
+  /* TODO: make the characters here configurable. */
+  if (memchr(buffer->str + len_before_value, self->delimiter, buffer->len - len_before_value) != NULL)
+    {
+      ScratchBuffersMarker marker;
+      GString *value_buffer = scratch_buffers_alloc_and_mark(&marker);
+
+      g_string_assign(value_buffer, buffer->str + len_before_value);
+      g_string_truncate(buffer, len_before_value);
+      g_string_append_c(buffer, '"');
+      append_unsafe_utf8_as_escaped_binary(buffer, value_buffer->str, value_buffer->len, "\"");
+      g_string_append_c(buffer, '"');
+
+      scratch_buffers_reclaim_marked(marker);
+    }
+
+  return TRUE;
+}
+
+static FilterXObject *
+_eval(FilterXExpr *s)
+{
+  FilterXFunctionFormatCSV *self = (FilterXFunctionFormatCSV *) s;
+
+  FilterXObject *csv_data = filterx_expr_eval_typed(self->input);
+  if (!csv_data)
+    {
+      filterx_eval_push_error("Failed to evaluate input. " FILTERX_FUNC_FORMAT_CSV_USAGE, s, NULL);
+      return NULL;
+    }
+
+  gboolean success = FALSE;
+  GString *formatted = scratch_buffers_alloc();
+
+  if (filterx_object_is_type(csv_data, &FILTERX_TYPE_NAME(list)))
+    {
+      guint64 size;
+      if (!filterx_object_len(csv_data, &size))
+        return FALSE;
+
+      gpointer user_data[] = { self, formatted };
+      success = TRUE;
+      for (guint64 i = 0; i < size && success; i++)
+        {
+          FilterXObject *elt = filterx_list_get_subscript(csv_data, i);
+          success = _append_to_buffer(NULL, elt, user_data);
+          filterx_object_unref(elt);
+        }
+    }
```

Comment:
[optional] I'd extract the specializations to separate functions to make this a bit easier to follow.